internal/internal_task_handlers.go

```diff
@@ -904,7 +946,11 @@ ProcessEvents:
 			if err != nil {
 				return nil, err
 			}
-			if w.isWorkflowCompleted {
+
+			// Break the event processing loop if the workflow is completed except in replay mode.
+			// In replay mode we check for nondeterminism cases and
+			// breaking the loop causes missing events in respondEvents which then causes false positives or false negatives.
+			if w.isWorkflowCompleted && !isInReplay {
 				break ProcessEvents
 			}
```

Comment:
to stick my initial feelings in here, which **have not been confirmed** and could be totally off:
I _suspect_ this is what's actually passing the tests, not the other changes.  This will mean that after the workflow-func returns, the loop continues _and gets more tasks_ and that does... something.  But it seems fairly likely that it'll lead to a different set of events -> successful non-determinism check.

Which can pretty clearly catch some things, but I suspect it's not actually _correct_ and it'll misbehave in other cases that we may not have tests for.

The "do not compare end requests" change is particularly suspect here IMO, as it feels backwards: we _very much_ care if the end event sequences differ, but this change is filtering them out so we can't tell.

So: interesting!  I'm genuinely surprised this was so simple and passed all tests.  That might imply it's more correct than I think it is... but I have moderately strong doubts at the moment.

Reply:
I added comments in various places to explain what is happening because this essentially has dependency on everything happening underneath in client code. Hope the comments help a little bit to clarify things. 

In summary, this change was needed because `w.isWorkflowCompleted` becomes true while there's still stuff to do in the history such as other activities to schedule/complete. So in replay mode we shouldn't break the loop early. Without looking at remaining events it's impossible to know if there's non-determinism issues. 
On top of this change I also had to change the `isDecisionEvent` for replays because the `WorkflowExecutionCompleted` (and other workflow end state events) should be ignored. Otherwise you end up with an extra event inside `respondEvents`. Previously this wasn't an issue because loop terminated before it ever got to the final event in the history :)

Reply:
> There's never a corresponding replay decision for `WorkflowExecutionCompleted`

Ah, yeah, good point - that does make the filtering-out make a bit more sense.

the linked [ref](https://github.com/uber-go/cadence-client/blob/1fed70036ff0018915a4d35fba9104c5752d6c7d/internal/internal_event_handlers.go#L811) there though is something else entirely.  that means that when we get a pre-recorded "complete" event, it doesn't trigger any progress in the workflow func... because it shouldn't, it should have already returned when handling the same decision task.

Reply:
The linked ref is relevant in the replay scenario. It points that event handler doesn't do anything for `WorkflowExecutionCompleted` and the loop is not able to grab something [here](https://github.com/uber-go/cadence-client/blob/25c0a0996205e4b35b3fa763ea29a44b4365f3d8/internal/internal_task_handlers.go#L976) to append to `replayDecisions`. Therefore [this comparison](https://github.com/uber-go/cadence-client/blob/25c0a0996205e4b35b3fa763ea29a44b4365f3d8/internal/internal_task_handlers.go#L993) fails unless I use the new `isDecisionTaskForReplay`

Reply:
gonna have to dig into this a bit more closely 🤔 
there's a lot of blending of "history has done X" and "replay has done Y" in that, and I'm not sure if this would be a reasonable addition or not.  or if it would be more accurately modeled by e.g. having a state machine for the workflow as a whole (probably just completed -> sent)...